ubuntu_image/classic_builder.py

```diff
@@ -0,0 +1,605 @@
+            if part.role is StructureRole.mbr or part.type == 'bare':
+                continue
+            activate = False
+            if (volume.schema is VolumeSchema.mbr and
+                    part.role is StructureRole.system_boot):
+                activate = True
+            elif (volume.schema is VolumeSchema.gpt and
+                    part.role is StructureRole.system_data and
+                    part.name is None):
+                part.name = 'writable'
+            image.partition(part.offset, part.size, part.name, activate)
+        # Now since we're done, we need to do a second pass to copy the data
+        # and set all the partition types.  This needs to be done like this as
+        # libparted's commit() operation resets type GUIDs to defaults and
+        # clobbers things like hybrid MBR partitions.
+        part_id = 1
+        for i, part in enumerate(volume.structures):
+            image.copy_blob(volume.part_images[i],
+                            bs=image.sector_size,
+                            seek=part.offset // image.sector_size,
+                            count=ceil(part.size / image.sector_size),
+                            conv='notrunc')
+            if part.role is StructureRole.mbr or part.type == 'bare':
+                continue
+            image.set_parition_type(part_id, part.type)
+            part_id += 1
+        for value, dest in offset_writes:
+            # Decipher non-numeric offset_write values.
+            if isinstance(dest, tuple):
+                dest = part_offsets[dest[0]] + dest[1]
+            image.write_value_at_offset(value // image.sector_size, dest)
+
+    def make_disk(self):
+        # Based on the -o/--output and -O/--output-dir options, and the volumes
+        # in the gadget.yaml file, we can now calculate where the generated
+        # disk images should go.  We'll write them directly to the final
+        # destination so they don't have to be moved later.  Here is the
+        # option precedence:
+        #
+        # * The location specified by -o/--output;
+        # * <output_dir>/<volume_name>.img
+        # * <work_dir>/disk.img
+        #
+        # If -o was given and there are multiple volumes, we ignore it and
+        # act as if -O is in use.
+        disk_img = None
+        if self.output is not None:
+            if len(self.gadget.volumes) > 1:
+                _logger.warn('-o/--output ignored for multiple volumes')
+            else:
+                disk_img = self.output
+        if not disk_img:
+            os.makedirs(self.output_dir, exist_ok=True)
+        # Walk through all partitions and write them to the disk image at the
+        # lowest permissible offset.  We should not have any overlapping
+        # partitions, the parser should have already rejected such as invalid.
+        #
+        # XXX: The parser should sort these partitions for us in disk order as
```

Comment:
Nitpick: again, this should be a TODO instead of a note. Especially since it uses "should", it makes it ambiguous whether it's already doing the checking and we're not sure it works, or it's not doing the checking currently but should.

Reply:
I guess this is a copy-paste comment from the builder.py code that Barry wrote.